src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java

```diff
@@ -424,6 +427,46 @@ public static boolean isItemSimilar(@Nullable ItemStack item, @Nullable ItemStac
         }
     }
 
+    public static boolean compareItem(@Nullable ItemStack itemOne, @Nullable ItemStack itemTwo) {
+        if (itemOne == null && itemTwo == null) {
+            return true;
+        } else if (itemOne == null || itemTwo == null) {
+            return false;
+        }
+
+        boolean itemOneHasMeta = itemOne.hasItemMeta();
+        boolean itemTwoHasMeta = itemTwo.hasItemMeta();
+
+        if ((!itemOneHasMeta && itemTwoHasMeta) || (itemOneHasMeta && !itemTwoHasMeta)) {
+            return false;
+        }
+
+        if (!itemOneHasMeta && !itemTwoHasMeta) {
+            return itemOne.getType() == itemTwo.getType();
+        }
+
+        ItemMeta itemMetaOne = itemOne.getItemMeta();
+        ItemMeta itemMetaTwo = itemTwo.getItemMeta();
+        if (itemMetaOne == null && itemMetaTwo == null) {
+            return true;
+        } else if (itemMetaOne == null || itemMetaTwo == null) {
+            return false;
+        }
+
+        Optional<String> itemDataOne = Slimefun.getItemDataService().getItemData(itemMetaOne);
+        Optional<String> itemDataTwo = Slimefun.getItemDataService().getItemData(itemMetaTwo);
+        if (itemDataOne.isPresent() && itemDataOne.equals(itemDataTwo)) {
+            Optional<DistinctiveItem> distinctiveItemOne = getDistinctiveItem(itemDataOne.get());
+            Optional<DistinctiveItem> distinctiveItemTwo = getDistinctiveItem(itemDataTwo.get());
+            if (distinctiveItemOne.isPresent() && distinctiveItemTwo.isPresent()) {
+                return distinctiveItemOne.get().canStack(itemMetaOne, itemMetaTwo);
+            }
+            return true;
+        }
+
+        return itemOne.equals(itemTwo);
```

Comment:
Doesn't this leave us with performance problems again?

Reply:
The idea is that the last resort is Spigot's slow one (see below though). 
We check IDs and our own stuff before this.

Only vanilla items with itemmeta will go down here and right now, the main performance loss is from getDisplayName. Considering we won't do that on almost all items hitting this... I think this is good. There's no way to not compare the itemstacks but still... know they're equal